src/os/unix/wayland.rs

```diff
@@ -0,0 +1,1182 @@
+                        *pointer_fl.borrow_mut() = true;
+                    }
+                    if !*keyboard_fl.borrow() && capabilities.contains(Capability::Keyboard) {
+                        *keyboard_fl.borrow_mut() = true;
+                    }
+                    if !*touch_fl.borrow() && capabilities.contains(Capability::Touch) {
+                        *touch_fl.borrow_mut() = true;
+                    }
+                }
+            });
+        }
+
+        event_queue
+            .sync_roundtrip(&mut (), |_, _, _| {})
+            .map_err(|e| Error::WindowCreate(format!("Roundtrip failed: {:?}", e)))
+            .unwrap();
+
+        let ret = (
+            *keyboard_fl.borrow(),
+            *pointer_fl.borrow(),
+            *touch_fl.borrow(),
+        );
+
+        ret
+    }
+}
+
+struct WaylandInput {
+    kb_events: mpsc::Receiver<wayland_client::protocol::wl_keyboard::Event>,
+    pt_events: mpsc::Receiver<wayland_client::protocol::wl_pointer::Event>,
+    input_devs: (Main<WlKeyboard>, Main<WlPointer>),
+}
+
+impl WaylandInput {
+    fn new(seat: &Main<WlSeat>) -> Self {
+        let (keyboard, pointer, _touch) = (seat.get_keyboard(), seat.get_pointer(), ());
+
+        let (kb_sender, kb_receiver) = mpsc::sync_channel(1024);
+
+        keyboard.quick_assign(move |_, event, _| {
+            kb_sender.send(event).unwrap();
+        });
+
+        let (pt_sender, pt_receiver) = mpsc::sync_channel(1024);
+
+        pointer.quick_assign(move |_, event, _| {
+            pt_sender.send(event).unwrap();
+        });
+
+        Self {
+            kb_events: kb_receiver,
+            pt_events: pt_receiver,
+            input_devs: (keyboard, pointer),
+        }
+    }
+
+    fn get_pointer(&self) -> &Main<WlPointer> {
+        &self.input_devs.1
+    }
+
+    fn iter_keyboard_events(
+        &self,
+    ) -> std::sync::mpsc::TryIter<wayland_client::protocol::wl_keyboard::Event> {
+        self.kb_events.try_iter()
+    }
+
+    fn iter_pointer_events(
+        &self,
+    ) -> std::sync::mpsc::TryIter<wayland_client::protocol::wl_pointer::Event> {
+        self.pt_events.try_iter()
+    }
+}
+
+pub struct Window {
+    display: DisplayInfo,
+
+    width: i32,
+    height: i32,
+
+    scale: i32,
+    bg_color: u32,
+    scale_mode: ScaleMode,
+
+    mouse_x: f32,
+    mouse_y: f32,
+    scroll_x: f32,
+    scroll_y: f32,
+    buttons: [u8; 3],
+    prev_cursor: CursorStyle,
+
+    should_close: bool,
+    active: bool,
+
+    key_handler: KeyHandler,
+    //option because MaybeUninit's get_ref() is nightly-only
+    keymap: Option<xkb::keymap::Keymap>,
+    update_rate: UpdateRate,
+    menu_counter: MenuHandle,
+    menus: Vec<UnixMenu>,
+    input: WaylandInput,
+    resizable: bool,
+    //temporary buffer
+    buffer: Vec<u32>,
+    //configure, close
+    toplevel_info: (Rc<RefCell<Option<(i32, i32)>>>, Rc<RefCell<bool>>),
```

Comment:
Clippy complains about this complex type. Maybe you could define a new type, or split it up?

Reply:
Which would you suggest packing into a struct? - Id put the last ones from key_handler into an own struct.

Reply:
Clippy didnt complain for me - Ill just push my new changes.